agent/src/agent_win.cc

```diff
@@ -11,87 +11,383 @@
   if (pipename.empty())
     return;
 
   pipename_ = pipename;
-  CreatePipe(&hPipe_);
+
+  connections_.reserve(kNumPipeInstances);
+  for (int i = 0; i < kNumPipeInstances; ++i) {
+    connections_.emplace_back(
+        std::make_unique<Connection>(pipename_, handler(), i == 0));
+    if (!connections_.back()->IsValid()) {
+      Shutdown();
+      break;
+    }
+  }
 }
 
 AgentWin::~AgentWin() {
   Shutdown();
 }
 
-std::unique_ptr<Session> AgentWin::GetNextSession() {
-  std::unique_ptr<SessionWin> session;
+void AgentWin::HandleEvents() {
+  std::vector<HANDLE> wait_handles;
+  for (;;) {
+    // Wait on any pipe event to happen.
+    GetHandles(wait_handles);
+    if (wait_handles.size() < kNumPipeInstances) {
+      break;
+    }
 
-  // Wait for a client to connect to the pipe.
-  DWORD err = ConnectPipe(hPipe_);
-  if (err == ERROR_SUCCESS) {
-    // Create a session with the existing pipe handle.  The session now owns
-    // the handle.
-    session = std::make_unique<SessionWin>(hPipe_);
-    hPipe_ = INVALID_HANDLE_VALUE;
-
-    err = session->Init();
-    if (err == ERROR_SUCCESS) {
-      // Create a new pipe to accept a new client connection.
-      err = CreatePipe(&hPipe_);
+    DWORD index = WaitForMultipleObjects(
+        wait_handles.size(), wait_handles.data(), FALSE, INFINITE);
+    if (index == WAIT_FAILED) {
+      DWORD err = GetLastError();
+      break;
     }
 
+    index -= WAIT_OBJECT_0;
+    auto& connection = connections_[index];
+    bool was_listening = !connection->IsConnected();
+    DWORD err = connection->HandleEvent(wait_handles[index]);
     if (err != ERROR_SUCCESS) {
-      session.reset();
+      // If `connection` was not listening and there are more than
+      // kNumPipeInstances pipes, delete this connection.  Otherwise
+      // reset it so that it becomes a listener.
+      if (!was_listening && connections_.size() > kNumPipeInstances) {
+        connections_.erase(connections_.begin() + index);
+      } else {
+        err = connection->Reset(pipename_);
+      }
     }
-  }
 
-  return session;
+    // If `connection` was listening and is now connected, create a new
+    // one so that there are always kNumPipeInstances listening.
+    if (err == ERROR_SUCCESS && was_listening && connection->IsConnected()) {
+      connections_.emplace_back(
+          std::make_unique<Connection>(pipename_, handler(), false));
+    }
+  }
 }
 
 int AgentWin::Stop() {
   Shutdown();
   return AgentBase::Stop();
 }
 
-DWORD AgentWin::CreatePipe(HANDLE* handle) {
-  DWORD err = ERROR_SUCCESS;
-
-  *handle = CreateNamedPipeA(pipename_.c_str(),
-    PIPE_ACCESS_DUPLEX, PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_WAIT |
-        PIPE_REJECT_REMOTE_CLIENTS,
-    PIPE_UNLIMITED_INSTANCES, kBufferSize, kBufferSize, 0, nullptr);
-  if (*handle == INVALID_HANDLE_VALUE) {
-    err = GetLastError();
+void AgentWin::GetHandles(std::vector<HANDLE>& wait_handles) const {
+  wait_handles.reserve(connections_.size());
```

Comment:
If the calls here are flipped so `clear` is called before `reserve`, then that implies no needless copies are done in `reserve` right? Or is there some nuance I'm missing that makes this ordering necessary?